arangod/Cache/FrequencyBuffer.h

```diff
@@ -42,74 +42,78 @@ namespace cache {
   T _empty;
-
+  
+private:
+  
+  static size_t powerOf2(size_t capacity) {
+    // TODO maybe use https://graphics.stanford.edu/~seander/bithacks.html#RoundUpPowerOf2
+    size_t i = 0;
+    for (; (static_cast<size_t>(1) << i) < capacity; i++) {
+    }
+    return (static_cast<size_t>(1) << i);
+  }
+  
  public:
   //////////////////////////////////////////////////////////////////////////////
   /// @brief Initialize with the given capacity.
   //////////////////////////////////////////////////////////////////////////////
   explicit FrequencyBuffer(size_t capacity)
-      : _capacity(0),
-        _mask(0),
-        _buffer(nullptr),
+      : _capacity(powerOf2(capacity)),
+        _mask(_capacity - 1),
+        _buffer(_capacity),
         _cmp(),
         _empty() {
-    size_t i = 0;
-    for (; (static_cast<size_t>(1) << i) < capacity; i++) {
-    }
-    _capacity = (static_cast<size_t>(1) << i);
-    _mask = _capacity - 1;
-    _buffer.reset(new std::vector<T>(_capacity));
-    TRI_ASSERT(_buffer->capacity() == _capacity);
-    TRI_ASSERT(_buffer->size() == _capacity);
+    TRI_ASSERT(_buffer.capacity() == _capacity);
+    TRI_ASSERT(_buffer.size() == _capacity);
   }
 
   //////////////////////////////////////////////////////////////////////////////
   /// @brief Reports the hidden allocation size (not captured by sizeof).
   //////////////////////////////////////////////////////////////////////////////
   static size_t allocationSize(size_t capacity) {
-    return sizeof(std::vector<T>) + (capacity * sizeof(T));
+    return capacity * sizeof(T);
   }
 
   //////////////////////////////////////////////////////////////////////////////
   /// @brief Reports the memory usage in bytes.
   //////////////////////////////////////////////////////////////////////////////
   size_t memoryUsage() const {
-    return ((_capacity * sizeof(T)) + sizeof(FrequencyBuffer<T>) +
-            sizeof(std::vector<T>));
+    return ((_capacity * sizeof(T)) + sizeof(FrequencyBuffer<T>));
   }
 
   //////////////////////////////////////////////////////////////////////////////
   /// @brief Insert an individual event record.
   //////////////////////////////////////////////////////////////////////////////
   void insertRecord(T record) {
     // we do not care about the order in which threads insert their values
-    (*_buffer)[basics::SharedPRNG::rand() & _mask] = record;
+    _buffer[basics::SharedPRNG::rand() & _mask].store(record, std::memory_order_relaxed);
   }
   
   //////////////////////////////////////////////////////////////////////////////
   /// @brief Remove all occurrences of the specified event record.
   //////////////////////////////////////////////////////////////////////////////
   void purgeRecord(T record) {
     for (size_t i = 0; i < _capacity; i++) {
-      if (_cmp((*_buffer)[i], record)) {
-        (*_buffer)[i] = _empty;
+      if (_cmp(_buffer[i], record)) {
+        _buffer[i].store(_empty, std::memory_order_relaxed);
       }
```

Comment:
This should probably be a `compare_and_exchange`.

Reply:
@mpoeter Is this the main race condition you were concerned about?

Reply:
Actually I referred to the fact that all operations on the vector were non-atomic and not otherwise synchronized, which strictly speaking is a _data race_.
This particular case would in fact be a race condition, so a compare-and-exchange would be appropriate. However, I just checked and it seems that this function is never actually used.

Reply:
Ah, before @graetzer's changes. Okay.

As far as the non-usage ... that's troubling. It should be used in at least one place. Thanks for catching that!

Reply:
if you have time feel free to work in  this PR

Reply:
Since the comparison operator is customizable you cannot assume that the two values are identical, i.e, you cannot use record as expected value in the compare-and-swap (CAS). Instead, load the value from buffer and use that value in the comparison and the CAS. BTW: all these operation can use relaxed memory order.